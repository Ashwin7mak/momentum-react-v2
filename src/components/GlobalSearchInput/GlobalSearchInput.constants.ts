const CLASS_PREFIX = 'md-global-search-input';

const DEFAULTS = {};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
