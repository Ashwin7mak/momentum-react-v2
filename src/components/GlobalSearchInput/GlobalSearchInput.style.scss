.md-global-search-input-wrapper {
  position: relative;
  border-radius: 6.25rem;
  width: 22.5rem;
  background-color: var(--theme-background-secondary-normal);
  border: 1px solid transparent;
  box-sizing: border-box;
  height: 1.75rem;
  display: flex;
  justify-content: center;
  align-content: center;

  &.search-input-focus {
    border: 1px solid var(--theme-outline-focus-normal);
  }

  &:hover {
    background-color: var(--theme-background-secondary-hover);
  }

  .input-container {
    position: relative;
    flex-grow: 1;
  }

  input {
    border: 0;
    background-color: transparent;
    box-shadow: none;
    color: var(--theme-text-primary-normal);
    font-family: var(--md-globals-font-default);
    width: calc(100% - 2rem);
    position: absolute;
    left: 0;
    font-size: 0.875rem;
    height: 1.75rem;

    &:focus-visible {
      outline: none;
    }

    &::placeholder {
      color: var(--theme-text-secondary-normal);
    }
  }  

  .search-icon {
    margin-left: 1.125rem;
    margin-right: 0.5625rem;
    height: 1.625rem;
    align-content: center;
    flex-grow: 0;
  }

  .clear-icon {
    border: 0;
    background-color: transparent;
    cursor: pointer;
    position: absolute;
    right: 0.75rem;
    top: 0.25rem;
    color: var(--theme-text-secondary-normal);

    &:active {
      color: var(--theme-text-primary-normal);
    }
  }

  .search-context-container {
    display: inline-flex;
    justify-content: center;
    align-content: center;
    flex-direction: column;
    height: 1.625rem;

    p {
      box-sizing: border-box;
      background-color: var(--theme-button-primary-normal);
      color: var(--theme-text-primary-inverted);
      z-index: 1;
      border-radius: 0.5rem;
      padding: 0.125rem 0.0625rem 0.125rem 0.0625rem;
      margin-top: 0.0625rem;
      margin-right: 0.0625rem;
      font-size: 0.875rem;
      height: 1.5rem;
      margin: 0;
      line-height: 1.5rem;
    }
  }

  ::-webkit-search-cancel-button{
    -webkit-appearance: none;
    pointer-events: none;
  }
}
