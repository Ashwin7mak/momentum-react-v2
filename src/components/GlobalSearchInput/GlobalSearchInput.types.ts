import { CSSProperties, ReactNode } from 'react';
import { AriaSearchFieldProps } from '@react-types/searchfield';

export interface Props extends AriaSearchFieldProps {
  /**
   * Child components of this ButtonPill.
   */
  children?: ReactNode;

  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;

  /**
   * Whether the global search is searching or not
   */
  searching?: boolean;

  /**
   * Number of characters to highlight in bubble
   */
  numHighlighted?: number;
}
