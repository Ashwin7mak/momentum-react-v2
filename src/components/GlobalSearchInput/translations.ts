import { defineMessages } from 'react-intl';

export default defineMessages({
  placeholder: {
    id: 'GlobalSearchInput.placeholder',
    defaultMessage: 'Search, meet, and call',
  },
  from: {
    id: 'GlobalSearchInput.from',
    defaultMessage: 'From:',
  },
  fromLabelEmpty: {
    id: 'GlobalSearchInput.fromLabelEmpty',
    defaultMessage: 'Choose who to filter by',
  },
  fromLabelNonEmpty: {
    id: 'GlobalSearchInput.fromLabelNonEmpty',
    defaultMessage: 'Filtering by {value}',
  },
  in: {
    id: 'GlobalSearchInput.in',
    defaultMessage: 'In:',
  },
  inLabelEmpty: {
    id: 'GlobalSearchInput.inLabelEmpty',
    defaultMessage: 'Choose a space to filter by',
  },
  inLabelNonEmpty: {
    id: 'GlobalSearchInput.inLabelNonEmpty',
    defaultMessage: 'Searching in {value}',
  },
  with: {
    id: 'GlobalSearchInput.with',
    defaultMessage: 'With:',
  },
  withLabelEmpty: {
    id: 'GlobalSearchInput.withLabelEmpty',
    defaultMessage: 'Choose someone who is in this space with you',
  },
  withLabelNonEmpty: {
    id: 'GlobalSearchInput.withLabelNonEmpty',
    defaultMessage: 'Searching for spaces with {value}',
  },
});
