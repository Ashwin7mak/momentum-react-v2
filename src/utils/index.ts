export { PrimitiveConverter } from './component-conversions';
