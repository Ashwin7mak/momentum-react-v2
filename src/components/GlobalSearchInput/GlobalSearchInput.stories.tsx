import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import GlobalSearchInput, { GlobalSearchInputProps } from './';
import argTypes from './GlobalSearchInput.stories.args';
import Documentation from './GlobalSearchInput.stories.docs.mdx';

export default {
  title: 'Momentum UI/GlobalSearchInput',
  component: GlobalSearchInput,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
  args: {
    // Args provided to all stories by default.
    placeholder: 'Search, meet, and call',
  },
};

/**
 * Primary story. This renders a single component with all external props.
 */
const Example = Template<GlobalSearchInputProps>(GlobalSearchInput).bind({});

Example.argTypes = { ...argTypes };

// TODO: Inject additional stories here.

/**
 * Common variants story. This renders multiple variants of a single component.
 */
const Common = MultiTemplate<GlobalSearchInputProps>(GlobalSearchInput).bind({});

Common.argTypes = { ...argTypes };
delete Common.argTypes.children;

Common.parameters = {
  variants: [
    { value: 'Typed' },
    { searching: true, placeholder: 'Searching...' },
    { disabled: true },
    { searchContext: 'With:' },
    { searchContext: 'From:' },
    { searchContext: 'In:' },
  ],
};

export { Example, Common };
