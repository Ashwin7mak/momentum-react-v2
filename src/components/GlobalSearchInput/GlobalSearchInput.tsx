/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { FC, useRef, useState } from 'react';
import classnames from 'classnames';
import { useFocus } from '@react-aria/interactions';

import ButtonSimple from '../ButtonSimple';
import { STYLE } from './GlobalSearchInput.constants';
import { Props } from './GlobalSearchInput.types';
import './GlobalSearchInput.style.scss';
import { useSearchField } from '@react-aria/searchfield';
import { useSearchFieldState } from '@react-stately/searchfield';
import { useIntl } from 'react-intl';

import Icon from '../Icon';
import messages from './translations';
/**
 * Global search input. Used for global search only
 */
const GlobalSearchInput: FC<Props> = (props: Props) => {
  const { className, id, style, searching, onKeyDown, onFiltersChange, filters = [] } = props;
  const [focus, setFocus] = useState(false);
  const { formatMessage } = useIntl();
  const state = useSearchFieldState(props);
  const ref = useRef(null);
  const { focusProps } = useFocus({
    onFocus: () => {
      setFocus(true);
    },
    onBlur: () => {
      setFocus(false);
    },
  });

  const { inputProps, clearButtonProps, labelProps } = useSearchField(props, state, ref);

  const additionalClasses = [];
  if (focus) {
    additionalClasses.push('search-input-focus');
  }

  const handleClick = () => {
    if (ref.current) {
      ref.current.focus();
    }
  };

  const handleKeyDown = (e) => {
    const { key } = e;
    if (key === 'Backspace') {
      onFiltersChange &&
        filters.length &&
        !e.target.value &&
        onFiltersChange(filters.slice(0, filters.length - 1));
    } else {
      if (onKeyDown) {
        onKeyDown(e);
      }
      inputProps.onKeyDown(e);
    }
  };

  const buildFilters = () => {
    const filterArray = [];
    const labels = [];
    filters.forEach((filter, index) => {
      labels.push(
        formatMessage(messages[`${filter.term}Label${filter.value ? 'Non' : ''}Empty`], {
          value: filter.value,
        })
      );

      const removePadding = index !== filters.length - 1;
      filterArray.push(
        <div
          key={filter.term}
          className={`search-context-container ${removePadding ? 'remove-padding' : ''}`}
        >
          <p>{`${formatMessage(messages[filter.term])} ${filter.value}`}</p>
        </div>
      );
    });

    const label = labels.length ? labels.join('. ') : formatMessage(messages.placeholder);

    return { filterArray, label };
  };

  const { label, filterArray } = buildFilters();

  return (
    <div
      className={classnames(className, STYLE.wrapper, ...additionalClasses)}
      id={id}
      onClick={handleClick}
      style={style}
    >
      <label htmlFor={inputProps.id} {...labelProps}>
        {label}
      </label>
      <Icon
        weight="light"
        scale={18}
        className="search-icon"
        name={searching ? 'spinner' : 'search'}
      />
      <div className="input-container">
        {!!filters.length && filterArray}
        <input {...inputProps} {...focusProps} ref={ref} onKeyDown={handleKeyDown} />
      </div>
      {state.value && (
        <ButtonSimple className="clear-icon" {...clearButtonProps}>
          <Icon scale={18} name="cancel" />
        </ButtonSimple>
      )}
    </div>
  );
};

export default GlobalSearchInput;
