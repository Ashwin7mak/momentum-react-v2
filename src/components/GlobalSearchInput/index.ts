import { default as GlobalSearchInput } from './GlobalSearchInput';
import * as CONSTANTS from './GlobalSearchInput.constants';
import { Props } from './GlobalSearchInput.types';

export { CONSTANTS as GLOBAL_SEARCH_INPUT_CONSTANTS };

export type GlobalSearchInputProps = Props;

export default GlobalSearchInput;
