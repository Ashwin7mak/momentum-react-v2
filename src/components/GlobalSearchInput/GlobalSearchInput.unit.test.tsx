import React from 'react';
import { mount } from 'enzyme';
import { SSRProvider } from '@react-aria/ssr';

import { mountAndWait } from '../../../test/utils';
import GlobalSearchInput, { GLOBAL_SEARCH_INPUT_CONSTANTS as CONSTANTS } from './';

describe('<GlobalSearchInput />', () => {
  describe('snapshot', () => {
    const mountComponent = (component) => {
      const container = mount(<SSRProvider>{component}</SSRProvider>);
      return container;
    };

    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mountComponent(<GlobalSearchInput aria-label="global search" />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mountComponent(
        <GlobalSearchInput aria-label="global search" className={className} />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mountComponent(<GlobalSearchInput aria-label="global search" id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mountComponent(
        <GlobalSearchInput aria-label="global search" style={style} />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when searching', () => {
      expect.assertions(1);

      const container = mountComponent(
        <GlobalSearchInput aria-label="global search" searching={true} />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot when numHighlighted is set', () => {
      expect.assertions(1);

      const container = mountComponent(
        <GlobalSearchInput aria-label="global search" value="From: someone" numHighlighted={5} />
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class', async () => {
      expect.assertions(1);

      const element = (await mountAndWait(<GlobalSearchInput aria-label="global search" />))
        .find(GlobalSearchInput)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', async () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = (
        await mountAndWait(<GlobalSearchInput aria-label="global search" className={className} />)
      )
        .find(GlobalSearchInput)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', async () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = (await mountAndWait(<GlobalSearchInput aria-label="global search" id={id} />))
        .find(GlobalSearchInput)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', async () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = (
        await mountAndWait(<GlobalSearchInput aria-label="global search" style={style} />)
      )
        .find(GlobalSearchInput)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    it('should pass the aria label to the input', async () => {
      expect.assertions(1);

      const element = (
        await mountAndWait(<GlobalSearchInput aria-label="global search" searching={true} />)
      )
        .find('input')
        .getDOMNode();

      expect(element.getAttribute('aria-label')).toBe('global search');
    });
  });

  describe('actions', () => {
    /* ...action tests... */
  });
});
