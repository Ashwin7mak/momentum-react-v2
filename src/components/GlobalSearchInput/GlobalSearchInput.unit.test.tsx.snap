// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalSearchInput /> snapshot should match snapshot 1`] = `
<SSRProvider>
  <GlobalSearchInput
    aria-label="global search"
  >
    <div
      className="md-global-search-input-wrapper"
      onClick={[Function]}
    >
      <Icon
        className="search-icon"
        name="search"
        scale={18}
        weight="light"
      />
      <div
        className="input-container"
      >
        <input
          aria-label="global search"
          disabled={false}
          id="react-aria-0-1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="search"
          value=""
        />
      </div>
    </div>
  </GlobalSearchInput>
</SSRProvider>
`;

exports[`<GlobalSearchInput /> snapshot should match snapshot when numHighlighted is set 1`] = `
<SSRProvider>
  <GlobalSearchInput
    aria-label="global search"
    numHighlighted={5}
    value="From: someone"
  >
    <div
      className="md-global-search-input-wrapper"
      onClick={[Function]}
    >
      <Icon
        className="search-icon"
        name="search"
        scale={18}
        weight="light"
      />
      <div
        className="input-container"
      >
        <input
          aria-label="global search"
          disabled={false}
          id="react-aria-0-1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="search"
          value="From: someone"
        />
        <div
          className="search-context-container"
        >
          <p
            aria-hidden="true"
          >
            From:
          </p>
        </div>
      </div>
      <ButtonSimple
        aria-label="Clear search"
        className="clear-icon"
        excludeFromTabOrder={true}
        onPress={[Function]}
        onPressStart={[Function]}
        preventFocusOnPress={true}
      >
        <FocusRing>
          <FocusRing
            focusClass="children"
            focusRingClass="md-focus-ring-wrapper children"
          >
            <button
              aria-label="Clear search"
              className="clear-icon"
              onBlur={[Function]}
              onClick={[Function]}
              onDragStart={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <Icon
                name="cancel"
                scale={18}
              />
            </button>
          </FocusRing>
        </FocusRing>
      </ButtonSimple>
    </div>
  </GlobalSearchInput>
</SSRProvider>
`;

exports[`<GlobalSearchInput /> snapshot should match snapshot when searching 1`] = `
<SSRProvider>
  <GlobalSearchInput
    aria-label="global search"
    searching={true}
  >
    <div
      className="md-global-search-input-wrapper"
      onClick={[Function]}
    >
      <Icon
        className="search-icon"
        name="spinner"
        scale={18}
        weight="light"
      />
      <div
        className="input-container"
      >
        <input
          aria-label="global search"
          disabled={false}
          id="react-aria-0-1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="search"
          value=""
        />
      </div>
    </div>
  </GlobalSearchInput>
</SSRProvider>
`;

exports[`<GlobalSearchInput /> snapshot should match snapshot with className 1`] = `
<SSRProvider>
  <GlobalSearchInput
    aria-label="global search"
    className="example-class"
  >
    <div
      className="example-class md-global-search-input-wrapper"
      onClick={[Function]}
    >
      <Icon
        className="search-icon"
        name="search"
        scale={18}
        weight="light"
      />
      <div
        className="input-container"
      >
        <input
          aria-label="global search"
          disabled={false}
          id="react-aria-0-1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="search"
          value=""
        />
      </div>
    </div>
  </GlobalSearchInput>
</SSRProvider>
`;

exports[`<GlobalSearchInput /> snapshot should match snapshot with id 1`] = `
<SSRProvider>
  <GlobalSearchInput
    aria-label="global search"
    id="example-id"
  >
    <div
      className="md-global-search-input-wrapper"
      id="example-id"
      onClick={[Function]}
    >
      <Icon
        className="search-icon"
        name="search"
        scale={18}
        weight="light"
      />
      <div
        className="input-container"
      >
        <input
          aria-label="global search"
          disabled={false}
          id="example-id"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="search"
          value=""
        />
      </div>
    </div>
  </GlobalSearchInput>
</SSRProvider>
`;

exports[`<GlobalSearchInput /> snapshot should match snapshot with style 1`] = `
<SSRProvider>
  <GlobalSearchInput
    aria-label="global search"
    style={
      Object {
        "color": "pink",
      }
    }
  >
    <div
      className="md-global-search-input-wrapper"
      onClick={[Function]}
      style={
        Object {
          "color": "pink",
        }
      }
    >
      <Icon
        className="search-icon"
        name="search"
        scale={18}
        weight="light"
      />
      <div
        className="input-container"
      >
        <input
          aria-label="global search"
          disabled={false}
          id="react-aria-0-1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="search"
          value=""
        />
      </div>
    </div>
  </GlobalSearchInput>
</SSRProvider>
`;
